relay-server/src/actors/project_cache.rs

```diff
@@ -625,36 +623,55 @@ impl ProjectCacheBroker {
     }
 
     /// Handles the processing of the provided envelope.
-    fn handle_processing(
-        &mut self,
-        state: Arc<ProjectState>,
-        sampling_state: Option<Arc<ProjectState>>,
-        envelope: Box<Envelope>,
-        envelope_context: EnvelopeContext,
-    ) {
+    ///
+    /// The following pre-conditions must be met before calling this function:
+    /// - Envelope's project state must be cached and valid.
+    /// - Optional: if dynamic sampling keys exists, the dyanmic project state must be cached and valid.
+    ///
+    /// Calling this function without envelope's project state available will cause the envelope to
+    /// be dropped and outcome will be logged.
+    fn handle_processing(&mut self, envelope: Box<Envelope>, envelope_context: EnvelopeContext) {
         let project_key = envelope.meta().public_key();
+
         // The `Envelope` and `EnvelopeContext` will be dropped if the `Project::check_envelope()`
         // function returns any error, which will also be ignored here.
-        let Some(Ok(checked)) = self
-            .projects
-            .get_mut(&project_key)
-            .map(|p| p.check_envelope(envelope, envelope_context)) else { return; };
-
-        if let Some((envelope, envelope_context)) = checked.envelope {
-            let mut process = ProcessEnvelope {
-                envelope,
-                envelope_context,
-                project_state: state.clone(),
-                sampling_project_state: None,
-            };
-
-            if let Some(sampling_state) = sampling_state {
-                if state.organization_id == sampling_state.organization_id {
-                    process.sampling_project_state = Some(sampling_state)
+        if let Some(project) = self.projects.get_mut(&project_key) {
```

Comment:
Personally I would write this as a [guard](https://en.wikipedia.org/wiki/Guard_(computer_science)) to keep the error handling close to the check, i.e.

```rust
let Some(project) = self.projects.get_mut(&project_key) else {
  // log error
  return
};

let Some(own_project_state) = ... else {
   // log error
   return
}

if let Ok(CheckedEnvelope {...}) = ... {
    ...
}